TASK: Radical cleanup for subscription pr to simplify and get into 9.0

mhsdesign opened the pull request:
Tries to simplify https://github.com/neos/neos-development-collection/pull/5321 to fit our current needs.
All the removed code was untested and was planned to be an extension point at some time. But to get Neos 9 ready i want to discuss if we can merge the foundation and revert this change later and introduce it well tested for Neos 9.1 !

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: Remove generic `EventHandlerInterface` for now, everything is a projection (can be later reintroduced)

mhsdesign committed:
TASK: Remove other generic subscription concepts not required for projections

- retry strategy
- runmode
- pausing, finishing
- InMemorySubscriptionStore

mhsdesign committed:
TASK: Remove `$skipBooting` because its an odd signature and unused

mhsdesign committed:
TASK: Remove removal of subscriptions

Detached subscriptions cannot be removed for now

mhsdesign committed:
TASK: Remove sqlite support for `DoctrineSubscriptionStore` as db locking is untested there

mhsdesign committed:
TASK: Remove subscription groups and filtering except for status and ids

Followup for remove other generic subscription concepts not required for projections

mhsdesign commented:
As discussed with @kitsunet we like the simplification for now and will also simplify this structure  

```php
final class Subscriber
{
    public function __construct(
        public readonly SubscriptionId $id,
        public readonly EventHandlerInterface $handler,
    ) {
    }
}

interface EventHandlerInterface
{
    // Some handlers might not need an explicit setup – for those this method can just be a no-op
    public function setup(): void;
    public function onBeforeCatchUp(SubscriptionStatus $subscriptionStatus): void;
    public function handle(EventInterface $event, EventEnvelope $eventEnvelope): void;
    public function onAfterCatchUp(): void;
}


final readonly class ProjectionEventHandler implements EventHandlerInterface
{
    /**
     * @param ProjectionInterface<ProjectionStateInterface> $projection
     */
    private function __construct(
        // publicly exposed for special cases
        public ProjectionInterface $projection,
        private CatchUpHookInterface|null $catchUpHook,
    ) {
    }

    // generic logic of EventHandlerInterface for setup, onBeforeCatchUp, handle, onAfterCatchUp
}
```

To better reflect the projection usecase by introducing a one and only `ProjectionSubscriber`. The reasoning for this is that having some logic on the handler and other on the public accessible projection of the handler is a complexity we dont need at this point. The idea is that all parts will be internal after all and only wired together by the content repository factory itself.

```php
final class ProjectionSubscriber
{
    /**
     * @param ProjectionInterface<ProjectionStateInterface> $projection
     */
    public function __construct(
        public readonly SubscriptionId $id,
        public readonly ProjectionInterface $projection,
        private readonly ?CatchUpHookInterface $catchUpHook
    ) {
    }

    public function onBeforeCatchUp(SubscriptionStatus $subscriptionStatus): void
    {
        $this->catchUpHook?->onBeforeCatchUp($subscriptionStatus);
    }

    public function handle(EventInterface $event, EventEnvelope $eventEnvelope): void
    {
        $this->catchUpHook?->onBeforeEvent($event, $eventEnvelope);
        $this->projection->apply($event, $eventEnvelope);
        $this->catchUpHook?->onAfterEvent($event, $eventEnvelope);
    }

    public function onAfterCatchUp(): void
    {
        $this->catchUpHook?->onAfterCatchUp();
    }
}
```

Future extensibility will be possibly by introducing another subscriber class and using that for example as union or with a common marker interface. It could similar but would be also different:

```php
final class SimpleSubscriber
{
    public function __construct(
        public readonly SubscriptionId $id,
        private EventHandlerInterface $handler,
    ) {
    }

    public function handle(EventInterface $event, EventEnvelope $eventEnvelope): void
    {
        $this->handler->handle($event, $eventEnvelope);
    }
}
```

We could even go back to the first draft but for now to keep it simple there will be just the one thing :)

mhsdesign committed:
TASK: Inline `ProjectionEventHandler` to `Subscriber` and make it a `SubscriberForProjection`

Simplification to make the projection use case first level
See https://github.com/neos/neos-development-collection/pull/5375#issuecomment-2491375655

bwaidelich reviewed (commented):
Thanks for creating a separate PR for this, that makes it easier to review & discuss!
In general, I totally agree to this simplification, i.e. removing subscription groups, run modes and retry strategy and removing unfinished status/engine interactions.

One little change that I would suggest: keep the naming "subscriber" vs "projection", because:
- there's a mismatch between variable name "additionalProjectionsFactory" and type "ContentRepositorySubsciberFactory"
- the distinction between projection and subscription makes sense even if the only supported type of subscription target projections (for now)

mhsdesign committed:
TASK: Remove `retry_attempt` from subscriptions

https://github.com/neos/neos-development-collection/pull/5321#issuecomment-2492882147

> Anyways, I think with the removed retry strategy we should just get rid of the automatic retry altogether right now.
  It's quite unlikely that a retry suddenly works without other changes. So I'd be fully OK if it was only possible to manually retry failed subscriptions for 9.0

mhsdesign committed:
TASK: Rename factory back to `$additionalSubscriberFactories`

> there's a mismatch between variable name "additionalProjectionsFactory" and type "ContentRepositorySubsciberFactory"
> the distinction between projection and subscription makes sense even if the only supported type of subscription target projections (for now)

https://github.com/neos/neos-development-collection/pull/5375#pullrequestreview-2453314204

mhsdesign commented:
Thank you for your response :) Im glad you like this 😅 Also good call to fully remove the retry attempt (had it hardcodeded first but yeah) And also it makes sense to keep the `ContentRepositorySubsciberFactory` naming and wiring as is.
(At first id didnt want that as i thought we would expose the `Subscriber` but this is just build internally in `build` and only the constructor of the factory is API so all good :D)
While testing i noticed that catchup hooks will not be registered for other projections, i have a stash to fix that but that will happen in the other pr.
So this task of removing obsolete stuff would be solved, onto the other tasks. Ill merge this, and maybe open a new pr when i do bigger changes again.
Have a nice vacation 🌞 🌞 🌞

mhsdesign committed:
BUGFIX: `discoverDetachedSubscriptions` did not persist changes

which lead to it being invoked on aktive

>  Subscriber "Vendor.Package:FakeProjection" could not invoke onBeforeCatchUp: Subscriber with the subscription id "Vendor.Package:FakeProjection" not found.
